### PR TITLE
Symbolic implementation of Index supporting tuple of slices.

### PR DESCRIPTION
**Pull request by jamesr66a**
No support for ellipses yet.

Tested on fairseq-py workload. Integration tests can be found at https://github.com/ezyang/onnx-pytorch/pull/24

### CODE REVIEW

File: torch/autograd/_functions/tensor.py [hunk 1]
```diff
@@ -24,13 +24,66 @@ def symbolic(g, i, index):
...
+            def append_advanced_index(index, dim, list, fillval):
+                if index is None:
+                    list.append(fillval)
+                else:
+                    addend = (dim if index < 0 else 0)
+                    list.append(index + addend)
+
+            for idx in range(ndims):
+                # Don't slice if we're off the end of the tuple
+                if idx > len(index):
```
Comment (houseroad) [inline torch/autograd/_functions/tensor.py]: The length of starts and ends should be the same as axes. In this case, the loop should end.

File: torch/autograd/_functions/tensor.py [hunk 2]
```diff
@@ -24,13 +24,66 @@ def symbolic(g, i, index):
...
+            # 3) positive index - append to list
+            # 4) integer - keep only that integer and squeeze it at the end
+            def append_advanced_index(index, dim, list, fillval):
+                if index is None:
+                    list.append(fillval)
+                else:
+                    addend = (dim if index < 0 else 0)
+                    list.append(index + addend)
+
+            for idx in range(ndims):
```
Comment (houseroad) [inline torch/autograd/_functions/tensor.py]: Probably, we should change ndims to len(index)

File: torch/autograd/_functions/tensor.py [hunk 3]
```diff
@@ -24,13 +24,66 @@ def symbolic(g, i, index):
...
+        if isinstance(index, int):
+            axes = g.constant(0, [1], "int")
+            starts = g.constant(index, [1], "long")
+            ends = g.constant(index + 1, [1], "long")
+            slice = g.op("Slice", i, axes, starts, ends)
+            return g.op("Squeeze", slice, axes_i=[0])
+        elif isinstance(index, tuple):
+            dims = i.type().sizes()
+            ndims = len(dims)
+            axes_ten = torch.IntTensor([idx for idx in range(ndims)])
```
Comment (houseroad) [inline torch/autograd/_functions/tensor.py]: Actually, range(ndims) ==> range(len(index))

**Review by houseroad — COMMENTED** [review]
Thanks! It looks good to me. 

One optimization:
We should only include len(index) dimensions in the "Slice" node.

File: torch/autograd/_functions/tensor.py [hunk 4]
```diff
@@ -24,13 +24,66 @@ def symbolic(g, i, index):
...
+            ends_list = []
+            squeeze_indices = []
+
+            # Given an index, size of dimension, a list, and a default fill val,
+            # fill in based on these conditions:
+            # 1) not specified (None) - fill with fillval (e.g. 0 or size)
+            # 2) negative index - calculate corresponding positive index and append
+            # 3) positive index - append to list
+            # 4) integer - keep only that integer and squeeze it at the end
+            def append_advanced_index(index, dim, list, fillval):
```
Comment (killeent) [inline torch/autograd/_functions/tensor.py]: Can we maybe name this something else? Its unclear to me how this relates to advanced indexing.

File: torch/autograd/_functions/tensor.py [hunk 5]
```diff
@@ -24,13 +24,66 @@ def symbolic(g, i, index):
...
+                append_advanced_index(index[idx].stop, dims[idx], ends_list, dims[idx])
+
+                if index[idx].step is not None:
+                    raise ValueError("Strided slice is not supported at this time")
+
+            starts_ten = torch.LongTensor(starts_list)
+            starts = g.op("Constant", value_t=starts_ten)
+            ends_ten = torch.LongTensor(ends_list)
+            ends = g.op("Constant", value_t=ends_ten)
+            slice = g.op("Slice", i, axes, starts, ends)
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: Please don't call variables `slice`. It's a Python builtin that lets you build slices outside of indexing operations, and you will shadow it

File: torch/autograd/_functions/tensor.py [hunk 6]
```diff
@@ -24,13 +24,66 @@ def symbolic(g, i, index):
...
+                    list.append(index + addend)
+
+            for idx in range(ndims):
+                # Don't slice if we're off the end of the tuple
+                if idx > len(index):
+                    starts_list.append(0)
+                    ends_list.append(dims[idx])
+                    continue
+
+                if isinstance(index[idx], int):
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: In Python2 there is also `long` that might appear there

File: torch/autograd/_functions/tensor.py (hunk 4 above)
Comment (apaszke) [reply torch/autograd/_functions/tensor.py]: Also, don't call arguments `list`. You're shadowing the `list ` constructor

**Review by apaszke — CHANGES_REQUESTED** [review]

**Pull request merged**